### PR TITLE
Players who edited a circuit assembly can scan it as a ghost.

### PR DESCRIPTION
**Pull request by ghost**
When you examine circuits you built after dying you get a link that shows the circuit's code.

:cl:
add: Players who edited a circuit assembly can scan it as a ghost.
/:cl:

[why]: So you can get your circuit's code if you die before finishing it.

I submitted a PR like this before but in that all players could scan the circuit even if they didn't make it so I closed that one.

### CODE REVIEW

**Comment by Tlaltecuhtli** [comment]
woah dude 🍆

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 1]
```diff
@@ -62,6 +63,9 @@
 	else
 		to_chat(user, "<span class='notice'>The maintainence panel [opened ? "can be" : "is"] <b>screwed</b> in place.</span>")
 
+	if(isobserver(user) && ckeys_allowed_to_scan.Find(user.ckey))
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/assemblies.dm]: `if(isobserver(user) && ckeys_allowed_to_scan[user.ckey])`

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 2]
```diff
@@ -181,6 +185,15 @@
 	if(..())
 		return 1
 
+	if(href_list["ghostscan"])
+		if(isobserver(usr) && ckeys_allowed_to_scan.Find(usr.ckey))
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/assemblies.dm]: Same

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 3]
```diff
@@ -264,6 +279,10 @@
 		to_chat(M, "<span class='notice'>The machine now has a label reading '[input]'.</span>")
 		name = input
 
+/obj/item/electronic_assembly/proc/add_allowed_scanner(ckey)
+	if(!ckeys_allowed_to_scan.Find(ckey))
+		ckeys_allowed_to_scan += ckey
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/assemblies.dm]: Replace two lines with `ckeys_allowed_to_scan[ckey] = TRUE`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by theo2003** [comment]
Fixed. Lists are maps in BYOND?

**Comment by Cyberboss** [comment]
Required reading: https://old.reddit.com/r/SS13/comments/5oplxp/what_is_the_main_problem_with_byond_as_an_engine/dclbu1a/

**Comment by theo2003** [comment]
Wow. BYOND is really messed up.

**Comment by Shirbu** [comment]
This is great, thank you theo.

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 4]
```diff
@@ -181,6 +185,15 @@
 	if(..())
 		return 1
 
+	if(href_list["ghostscan"])
+		if(isobserver(usr) && ckeys_allowed_to_scan[usr.ckey])
+			if(length(src.assembly_components))
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/assemblies.dm]: `if(assembly_components.len)`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**